r-lib-error-1.0.md

```diff
@@ -0,0 +1,60 @@
+
+# Specification of r-lib errors, version 1.0
+
+Note: this is an experimental draft.
+
+Note: version 1.0 corresponds to the `rlang_error` class, as
+implemented in rlang 0.4.0.
+
+## Classes
+
+An r-lib error object always has at least classes `rlib_error`, `error`
+and `condition`. It may have other classes.
+
+## Members
+
+An `rlib_error` object has the following members:
+
+* `message`: `character(1)`, the error message.
+* `trace`: optional, an `rlib_trace` object, see below. 
+* `parent`: optional, pointer to a parent error, of class `error`. It
+  may also be `NULL`, which is the same as missing.
+* `call`: optional, the captured call, of class `call`.
+* `version`: optional, `r_lib` error version, defaults to 1.0, i.e. this
+  specification.
+
+## Traces
+
+An `rlib_trace` object has the following members:
+
+* `calls`: A list of calls, each of class `call`. This is the stack when
+  the error was thrown. `calls[[1]]` is the top of the stack.
+* `parents`: Integer vector of parent frame pointers, as in `sys.parents()`.
+* `envs`: List or character vector of environment names. They are typically
+  package names for frames that correspond to functions in packages.
+* `indices`: Frame numbers, typically just 1:n.
+* `version`: options, `r_lib` error version, defaults to 1.0, i.e. this
+  specification.
+
+`calls`, `parents`, `envs` and `indices` must have the same length.
```

Comment:
I wish we could use a data frame here.

Reply:
Why can't we, actually? In the next version we can, now?

Reply:
In general it is cumbersome to use the data.frame class for low-level things, because of all the complex semantics and corner cases to think about.

Reply:
Isn't this the right place for a record? (Which we can't do because it would use vctrs, and vctrs needs to import rlang)